a3_stereo_image_proc benchmark

jasonjabbour opened the pull request:
As follows from #8, the a3_stereo_image_proc benchmark now reuses the image_input_component code from the a1 benchmark. Slight modifications were made to the image_input_component.hpp and image_input_component.cpp files to better accommodate the a3 benchmark. I also created a yaml file instead of a readme file.

jwdinius commented on benchmarks/perception/a1_perception_2nodes/src/image_input_component.cpp:
```diff
@@ -39,8 +39,12 @@ namespace perception
+
   // Create image pub
-  pub_image_ = image_transport::create_camera_publisher(this, "input");
+  pub_image_ = image_transport::create_camera_publisher(this, input_topic_name);
```
Is `input_topic_name` used anywhere else besides the node constructor?  If not, consider making `input_topic_name` locally scoped within the constructor as there is no need for a (protected) member.

jasonjabbour replied:
gotcha! No, it's not used anywhere else. I changed it to be locally scoped. Thanks!

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+cmake_minimum_required(VERSION 3.5)
```
Is this file based on a project-wide standard?  Is there a way to standardize boilerplate files like this across different benchmarks?

jasonjabbour replied:
@vmayoral I saw that you were using this version in your CMake file in the a1 and a2 benchmarks so I kept it consistent in the a3 benchmark. Is there any particular reason you decided to use this version?

jwdinius replied:
Sorry, this was meant to be a comment regarding the entire file (and standard files like it), not the line in the file.  I was asking if there was a benchmark template folder that includes boilerplate files like this.  I believe this is a question for @vmayoral.

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/benchmark.yaml:
```diff
@@ -0,0 +1,85 @@
+id: a3
```
One advantage of markdown over yaml in this context would be that GitHub beautifies markdown rendered in the browser.  Is there a similar beautifier for yaml?  How can the contents of this file be easily parsed by a reader/user?

vmayoral replied:
We've got tools that convert `yaml` into Markdown automatically. See https://github.com/robotperf/benchmarks/blob/main/benchmarks/other/ros2benchmark/ros2benchmark/verb/update.py.

This can be invoked as:

```bash
ros2 benchmark update
```

and should run in simple CI job in this repo after each commit. That way we keep `yaml` and `README.md` files consistent. For archiving/databases purposes, I'd vote for yaml or similar. Markdown's worse in my experience doing that (and actually not meant for it, though of course you could hack it for that).

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/benchmark.yaml:
```diff
@@ -0,0 +1,85 @@
+  ros2 run teleop_twist_keyboard teleop_twist_keyboard
+
+  # Verify: Play back rosbag after recording it in a loop
+  ros2 bag play --loop /path/to/bag/directory/
```
Are there guidelines around how complex the launch process should be for benchmarks?  Having to launch `N` (=4 or 5, in this case) is labor-intensive.  Couldn't a convenience script be provided for the end-user to spawn all processes at once?

jasonjabbour replied:
Yeah, that makes sense! I just added a script that will spawn all the optional processes at once!

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/config/empty.yaml:
```diff
@@ -0,0 +1 @@
+# This is an empty file, so that git commits the folder correctly
```
Why do you need this folder under version control if there is no meaningful content in it?

jasonjabbour replied:
Yeah, that was some old code, I'll delete it, thanks!

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/description/inertial_macros.xacro:
```diff
@@ -0,0 +1,40 @@
+    <!-- Specify some standard inertial calculations https://en.wikipedia.org/wiki/List_of_moments_of_inertia -->
+    <!-- These make use of xacro's mathematical functionality -->
+
+    <xacro:macro name="inertial_sphere" params="mass radius *origin">
```
The code for the simulated robot and sensor models might be useful for other benchmarks in the future.  Could this code be moved to a separate package that can be imported by other benchmarks?

jasonjabbour replied:
Resolving this issue due to our discussion in the latest HAWG meeting

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/include/a3_stereo_image_proc/disparity_output_component.hpp:
```diff
@@ -0,0 +1,46 @@
+#include <rclcpp/rclcpp.hpp>
+#include <image_transport/image_transport.hpp>
+#include <ament_index_cpp/get_resource.hpp>
+#include <stereo_msgs/msg/disparity_image.hpp>
```
Typically, `stdlib` and other system headers come first and project headers are last.  Is this convention standard for this project?

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/include/a3_stereo_image_proc/disparity_output_component.hpp:
```diff
@@ -0,0 +1,46 @@
+#include <sstream>
+#include <string>
+#include <vector>
+#include <thread>
```
From the context below, you don't need any of these headers included in this interface definition file.  Consider moving to the source implementation if they are really needed.

jasonjabbour replied:
thanks for pointing this out, yep, they aren't needed. removed!

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/launch/launch_sim.launch.py:
```diff
@@ -0,0 +1,46 @@
+    # Include the robot_state_publisher launch file, provided by our own package. Force sim time to be enabled
+    # !!! MAKE SURE YOU SET THE PACKAGE NAME CORRECTLY !!!
+
+    package_name='a3_stereo_image_proc' #<--- CHANGE ME
```
Does this need to be changed?

jasonjabbour replied:
nope! thanks for pointing this out! Made the change :)

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/launch/trace_a3_stereo_image_proc.launch.py:
```diff
@@ -0,0 +1,93 @@
+                'userspace': ['vpid', 'vtid', 'procname'],
+        },
+        # events_kernel=DEFAULT_EVENTS_KERNEL,
+        # context_names=DEFAULT_CONTEXT,
```
Why not remove the commented lines?

vmayoral replied:
+1 to this. My fault probably for leaving it in other examples.

jwdinius commented on benchmarks/perception/a3_stereo_image_proc/src/disparity_output_component.cpp:
```diff
@@ -0,0 +1,61 @@
+
+
+
+
```
Why all of these extra lines?

jasonjabbour replied:
oops thanks!!

vmayoral commented on benchmarks/perception/a3_stereo_image_proc/launch/trace_a3_stereo_image_proc.launch.py:
```diff
@@ -0,0 +1,93 @@
+                'kernel': [],
+                'userspace': ['vpid', 'vtid', 'procname'],
+        },
+        # events_kernel=DEFAULT_EVENTS_KERNEL,
```
```suggestion
```

vmayoral commented on benchmarks/perception/a3_stereo_image_proc/launch/trace_a3_stereo_image_proc.launch.py:
```diff
@@ -0,0 +1,93 @@
+                'userspace': ['vpid', 'vtid', 'procname'],
+        },
+        # events_kernel=DEFAULT_EVENTS_KERNEL,
+        # context_names=DEFAULT_CONTEXT,
```
```suggestion
```

jasonjabbour commented on benchmarks/perception/a1_perception_2nodes/include/a1_perception_2nodes/image_input_component.hpp:
```diff
@@ -25,6 +25,7 @@ class ImageInputComponent
   explicit ImageInputComponent(const rclcpp::NodeOptions &);
 
 protected:
+  std::string input_topic_name;
```
Just changed `input_topic_name` to be locally scoped. So in that case I'm not going to add a trailing `_`, right?

jwdinius replied:
Correct.  The trailing `_` is for class members only.

jasonjabbour commented on benchmarks/perception/a3_stereo_image_proc/src/disparity_output_component.cpp:
```diff
@@ -0,0 +1,55 @@
+void DisparityOutputComponent::disparityCb(
+  const stereo_msgs::msg::DisparityImage::SharedPtr disparity_msg)
+{
+  TRACEPOINT(
+    robotperf_image_output_cb_init,
+    "Initializing image output callback",
+    static_cast<const void *>(this),
+    static_cast<const void *>(&(*disparity_msg)));
```
@vmayoral, while I was trying to analyze the data I realized my output tracepoint might not be completely correct. I used `robotperf_image_output_cb_init` because I saw that in your A1 benchmark, but where does `robotperf_image_output_cb_init` come from and should it change if I am tracing a disparity message type?

jasonjabbour commented:
@vmayoral Analyzed the data and added the latency to the yaml file!

vmayoral commented:
@jasonjabbour I'll have time tomorrow to review this. Is it ready to be reproduced in other machines? including the analysis script?

jasonjabbour commented:
The analysis script doesn't terminate properly. Trying to fix it now ...

![image](https://user-images.githubusercontent.com/55008744/234090364-31369ec4-dc1d-47d4-8dbd-5716a195a6e0.png)

vmayoral commented:
> The analysis script doesn't terminate properly. Trying to fix it now ...
> 
> ![image](https://user-images.githubusercontent.com/55008744/234090364-31369ec4-dc1d-47d4-8dbd-5716a195a6e0.png)

Mmmm, from the output of this, are you sure you're finding any *matching* sets at all? Meaning, do you complete the chain at least once? Try printing the `len` of the resulting list (of lists) after the processing routine.

If you're not finding any matching sets at all, then there might be some issues in the processing of the sets you need to debug further.

I'll plan to have a look at this myself tomorrow, so feel free to describe in here your debugging process. I'll build on top of it.

jasonjabbour commented:
Using babeltrace, a preview of my data looks like this:

![image](https://user-images.githubusercontent.com/55008744/234093244-b16a6f99-aa0a-4c32-b93b-a050afd1fd96.png)

and I can identify these tracepoints in my data:

```
robotperf_benchmarks:robotperf_image_input_cb_init
robotperf_benchmarks:robotperf_image_input_cb_fini

robotperf_benchmarks:robotperf_image_input_cb_init
robotperf_benchmarks:robotperf_image_input_cb_fini

robotperf_benchmarks:robotperf_image_output_cb_init
robotperf_benchmarks:robotperf_image_output_cb_fini
```

So, I tried to modify the target chain (and the other lists below it)

![image](https://user-images.githubusercontent.com/55008744/234093561-8a539d5f-fa57-4bfd-bf54-eb4e833c8db8.png)

By printing the `image_pipeline_msg_sets` list of lists I get the following output and a length of 326:

![image](https://user-images.githubusercontent.com/55008744/234094221-3f838e47-84ed-4f21-b2e2-17845c2702b4.png)

I think I am finding matching sets:

![image](https://user-images.githubusercontent.com/55008744/234095443-ee5b1cd1-181b-4f80-95c5-4733d7eb3ebc.png)

I believe the error might be an issue within the `traces()` function

vmayoral commented:
@jasonjabbour been searching for the dataset and couldn't find it anywhere. Any chance you can facilitate it to reproduce results?

jasonjabbour commented:
@vmayoral Thanks for the help today! I cleaned up some very minor things and I added the max latency of running the benchmark on my Intel i9 CPU. I made a note of the other statistics in the yaml file. I also uploaded the data for this benchmark here: https://drive.google.com/drive/u/0/folders/1Z9JtllJd0X7aVcBZoYZ56WEHRPEbUB_N

I think it should be ready! Let me know if I should change something!

vmayoral commented:
Thanks for this @jasonjabbour, <ins>great contribution</ins>.

Tested locally (*Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz*):

 |   | Benchmark Mean | Benchmark RMS | Benchmark Max  | Benchmark Min | Mean | RMS | Max | Min |
 | --- | --- | --- | --- | --- | --- | --- | --- | --- |
 | RobotPerf benchmark | **29.10** ms | **29.22** ms | **34.94** ms | **6.36** ms | **43.88** ms | **44.33** ms | **65.14** ms | **10.80** ms |

jasonjabbour commented:
Awesome!! Thanks, Victor!
